xds/src/main/java/io/grpc/xds/XdsNameResolver.java

```diff
@@ -444,6 +555,177 @@ public void run() {
         });
       }
     }
+
+    @Nullable
+    private Long determineFaultDelayNanos(FaultDelay faultDelay, Metadata headers) {
+      Long delayNanos;
+      Integer delayRate = null;
+      if (faultDelay.headerDelay()) {
+        try {
+          int delayMillis = Integer.parseInt(headers.get(HEADER_DELAY_KEY));
+          delayNanos = TimeUnit.MILLISECONDS.toNanos(delayMillis);
+          String delayPercentageStr = headers.get(HEADER_DELAY_PERCENTAGE_KEY);
+          if (delayPercentageStr != null) {
+            int delayPercentage = Integer.parseInt(delayPercentageStr);
+            delayRate = delayPercentage * 100 * 100;
+          }
+        } catch (NumberFormatException e) {
+          return null; // treated as header_delay not applicable
+        }
+      } else {
+        delayNanos = faultDelay.delayNanos();
+      }
+      if (delayRate == null) {
+        delayRate = faultDelay.ratePerMillion();
+      }
+      if (random.nextInt(1_000_000) >= delayRate) {
+        return null;
+      }
+      return delayNanos;
+    }
+
+    @Nullable
+    private Status determineFaultAbortStatus(FaultAbort faultAbort, Metadata headers) {
+      Status abortStatus = null;
+      Integer abortRate = null;
+      if (faultAbort.headerAbort()) {
+        try {
+          String httpCodeStr = headers.get(HEADER_ABORT_HTTP_STATUS_KEY);
+          if (httpCodeStr != null) {
+            int httpCode = Integer.parseInt(httpCodeStr);
+            abortStatus = GrpcUtil.httpStatusToGrpcStatus(httpCode);
+          }
+          String grpcCodeStr = headers.get(HEADER_ABORT_GRPC_STATUS_KEY);
+          if (grpcCodeStr != null) {
+            int grpcCode = Integer.parseInt(grpcCodeStr);
+            abortStatus = Status.fromCodeValue(grpcCode);
+          }
+          String abortPercentageStr = headers.get(HEADER_ABORT_PERCENTAGE_KEY);
+          if (abortPercentageStr != null) {
+            int abortPercentage =
+                Integer.parseInt(headers.get(HEADER_ABORT_PERCENTAGE_KEY));
+            abortRate = abortPercentage * 100 * 100;
+          }
+        } catch (NumberFormatException e) {
+          return null; // treated as header_abort not applicable
+        }
+      } else {
+        abortStatus = faultAbort.status();
+      }
+      if (abortRate == null) {
+        abortRate = faultAbort.ratePerMillion();
+      }
+      if (random.nextInt(1_000_000) >= abortRate) {
+        return null;
+      }
+      return abortStatus;
+    }
+  }
+
+  /**
+   * A forwarding client call that counts active fault injections.
+   */
+  private final class ActiveFaultCountingClientCall<ReqT, RespT> extends
```

Comment:
I'd rather just put all fault injection logic into one ClientCall implementation rather than having multiple layers of wrapping. It's easy to do so as all of them are private. Wrapped calls are not debugging friendly.

Reply:
ActiveFaultCountingClientCall is a simple solution for three cases: delay-injected, abort-injected, delay-then-abort-injected. Putting all fault injection logic into one ClientCall implementation is easy to introduce bugs.